pageserver/src/tenant/storage_layer/image_layer.rs

```diff
@@ -605,9 +613,26 @@ impl ImageLayerInner {
             match res {
                 Ok(blobs_buf) => {
                     let frozen_buf = blobs_buf.buf.freeze();
-
                     for meta in blobs_buf.blobs.iter() {
                         let img_buf = frozen_buf.slice(meta.start..meta.end);
+                        let decompressed =
+                            vectored_blob_io::decompress(&img_buf, meta.compression_bits).await;
+
+                        let img_buf = match decompressed {
+                            Ok(None) => img_buf,
+                            Ok(Some(decompressed)) => decompressed,
+                            Err(e) => {
+                                reconstruct_state.on_key_error(
+                                    meta.meta.key,
+                                    PageReconstructError::Other(anyhow!(e).context(format!(
+                                        "Failed to decompress blob from virtual file {}",
+                                        self.file.path,
+                                    ))),
+                                );
```

Comment:
@VladLazar would there be any theoretical harm if we call reconstruct_state.on_key_error twice for the same key? I don't think it's possible in image layers because each image layer only has at most one entry per key. But still.

Reply:
No harm. It will clobber the previously recorded error.